### PR TITLE
spec clarification for matmulinteger and qlinearmatmul

### PR DESCRIPTION
**Pull request by askhade**
Signed-off-by: Ashwini Khade <askhade@microsoft.com>

**Description**
Clarify spec for zero point and scale shape expectations in case of ND inputs
- Describe your changes.

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: onnx/defs/math/defs.cc [hunk 1]
```diff
@@ -2343,9 +2343,11 @@ Matrix product that behaves like numpy.matmul: https://docs.scipy.org/doc/numpy-
 It consumes two quantized input tensors, their scales and zero points, scale and zero point of output, and computes the quantized output.
 The quantization formula is y = saturate((x / y_scale) + y_zero_point). For (x / y_scale), it is rounding to nearest ties to even.
 Refer to https://en.wikipedia.org/wiki/Rounding for details. Scale and zero point must have same shape.
-They must be either scalar (per tensor) or 1-D tensor (per row for 'a' and per column for 'b'). If scale and zero point are 1-D tensor,
-the number of elements of scale and zero point tensor of input 'a' and output 'y' should be equal to the number of rows of input 'a',
-and the number of elements of scale and zero point tensor of input 'b' should be equal to the number of columns of input 'b'.
+They must be either scalar (per tensor) or N-D tensor (per row for 'a' and per column for 'b'). Scalar refers to per tensor quantization 
+whereas N-D refers or per row or per column quantization. If the input is 2D of shape [M, K] then zero point and scale tensor may be an M element vector [v_1, v_2, ..., v_M]
+for per row quantization and K element vector of shape [v_1, v_2, v_K] for per column quantization.
```
Comment (yufenglee) [inline onnx/defs/math/defs.cc:2348]: nit: ...

File: onnx/defs/math/defs.cc [hunk 2]
```diff
@@ -2520,9 +2524,10 @@ ONNX_OPERATOR_SET_SCHEMA(
         .Input(
             3,
             "b_zero_point",
-            "Zero point tensor for input 'B'. It's optional and default value is 0.  It could be a scalar or a 1-D tensor, "
-            "which means a per-tensor or per-column quantization. If it's a 1-D tensor, its number "
-            "of elements should be equal to the number of columns of input 'B'.",
+            "Zero point tensor for input 'B'. It's optional and default value is 0. It could be a scalar or a N-D tensor, "
+            "Scalar refers to per tensor quantization whereas N-D refers or per col quaization. "
```
Comment (gramalingam) [inline onnx/defs/math/defs.cc:2528]: Typo: change to "to per column quantization"

File: onnx/defs/math/defs.cc [hunk 3]
```diff
@@ -2509,9 +2512,10 @@ ONNX_OPERATOR_SET_SCHEMA(
         .Input(
             2,
             "a_zero_point",
-            "Zero point tensor for input 'A'. It's optional and default value is 0. It could be a scalar or a 1-D tensor, "
-            "which means a per-tensor or per-row quantization. If it's a 1-D tensor, its number of elements "
-            "should be equal to the number of rows of input 'A'.",
+            "Zero point tensor for input 'A'. It's optional and default value is 0. It could be a scalar or N-D tensor. "
+            "Scalar refers to per tensor quantization whereas N-D refers or per row quaization. "
```
Comment (gramalingam) [inline onnx/defs/math/defs.cc:2516]: typo: change to "to per row quantization"

File: onnx/defs/math/defs.cc [hunk 4]
```diff
@@ -2509,9 +2512,10 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+            "If the input is 2D of shape [M, K] then zero point tensor may be an M element vector [zp_1, zp_2, ..., zp_M]. "
+            "If the input is N-D tensor with shape [D1, D2, M, K] then zero point tesnor may have shape [D1, D2, M, 1]. ",
```
Comment (gramalingam) [inline onnx/defs/math/defs.cc:2518]: But can it be [D1, D2, 1]? Or, do we allow only the two extremes of a zero-point for the entire tensor or a zero-point for each row?

Comment (gramalingam) [reply onnx/defs/math/defs.cc:2518]: Or [D1, D2, 1, 1], I mean.

File: onnx/defs/math/defs.cc [hunk 5]
```diff
@@ -2340,12 +2340,15 @@ ONNX_OPERATOR_SET_SCHEMA(
...
-the number of elements of scale and zero point tensor of input 'a' and output 'y' should be equal to the number of rows of input 'a',
-and the number of elements of scale and zero point tensor of input 'b' should be equal to the number of columns of input 'b'.
+It consumes two quantized input tensors, their scales and zero points, scale and zero point of output, 
+and computes the quantized output. The quantization formula is y = saturate((x / y_scale) + y_zero_point). 
+For (x / y_scale), it is rounding to nearest ties to even. Refer to https://en.wikipedia.org/wiki/Rounding for details. 
+Scale and zero point must have same shape. They must be either scalar (per tensor) or N-D tensor 
+(per row for 'a' and per column for 'b'). Scalar refers to per tensor quantization whereas N-D refers or per row 
+or per column quantization. If the input is 2D of shape [M, K] then zero point and scale tensor may be 
+an M element vector [v_1, v_2, ..., v_M] for per row quantization and K element vector of shape [v_1, v_2, ..., v_K] 
+for per column quantization. If the input is N-D tensor with shape [D1, D2, M, K] then zero point and scale tesnor may
```
Comment (gramalingam) [inline onnx/defs/math/defs.cc:2350]: type "tesnor"

File: onnx/defs/math/defs.cc (hunk 4 above)
Comment (askhade) [reply onnx/defs/math/defs.cc:2518]: right now we only allow per tensor or per row\col  zero points

**Review by yufenglee — APPROVED** [review]

**Review by gramalingam — APPROVED** [review]

**Pull request merged**